### PR TITLE
feat: switch Lease lifecycle to remote-lease stubs (#142)

### PR DESCRIPTION
**Pull request by gmanev7**
## Summary

Switch the Nolus lease contract's open-side lifecycle from the legacy ICA-open + DEX-side path to remote-lease controller stubs that bridge to a Solana counterparty via IBC. Lays the foundation for `nolus-protocol/ibc-solray#142`. Subsequent work (swap replacement, TransferOut, CloseLease, leaser v8 bump, dead-code removal) lands in follow-up PRs.

## Changes

- **New `OpenLease` state** (`protocol/contracts/lease/src/contract/state/opening/open_lease.rs`). `RequestLoan::on_response` now emits `Factory::open` to the configured `remote_lease_controller` and transitions to `OpenLease`. On `OperationOk` the lease threads the Solana-side `RemoteLeaseId` into the existing `super::buy_asset::start` cascade (additive — legacy `OpenIcaAccount` → `BuyAsset` path retained for now). On `OperationErr` / `OperationTimeout` an atomic refund batch repays the LPP loan, returns the downpayment, calls `LeasesRef::finalize_lease`, emits `wasm-ls-remote-lease-open-failed`, and transitions to the new `OpenFailed` terminal.
- **New `OpenFailed` terminal** (`protocol/contracts/lease/src/contract/state/open_failed.rs`). Authenticated late-ack absorber for UNORDERED-channel stale OK acks — emits `wasm-ls-remote-lease-late-ack`, no state mutation. Gated by `LeasesRef::remote_lease_callback_permission` even though idempotent.
- **`LeaseDTO.remote_lease_id: RemoteLeaseId`** (non-optional) persists the Solana PDA. Storage v9→v10 with refuse-migrate (`UnsupportedMigration`); mainnet v9-population is zero so the refuse is correct.
- **Query surface**: new `StateResponse::OpenFailed { reason }`. `OngoingTrx` reshaped — `OpenIcaAccount` unit variant dropped, `RequestingOpenLease` (pre-ack, no PDA known) and `OpenLease { remote_lease }` (post-ack) added.
- **Leaser `Leases::save`** returns `SaveOutcome::{Registered, AlreadyRegistered, Cancelled}`. New `CANCELLED_PENDING` Storage sentinel set by `Leases::remove` lets `save` distinguish a deliberate auto-refund cancellation (`Ok(Cancelled)`) from a missing `cache_open_req` programmer error (`Err(PendingCustomerNotCached)`). Address collisions surface as `DuplicateLeaseAddress` rather than a silent `debug_assert!`.
- **`NewLeaseContract`** gains `remote_lease_controller: Addr` + `expected_instance_ordinal: u16`, populated by `Borrow::open_lease_msg` from the leaser `Config`.
- **`OpenLeaseParams::new`** wire invariant relaxed: only `lpn_currency != asset_currency` is enforced. `downpayment` may overlap either of the other two. Cross-repo follow-up filed at `nolus-protocol/ibc-solray#331`.
- **Stand-in remote-lease controller** in `tests/src/common/remote_lease_controller_stub.rs` drives the three new integration tests.

## Verification

- Built and lint-clean across the affected crates (`cargo clippy --features=dex-osmosis --all-targets -- -D warnings` on lease, leaser, remote_lease, remote_lease_wire, and integration_tests).
- Unit tests pass: lease 88, leaser 33, remote_lease 42, remote_lease_wire 52.
- Integration tests pass: 119 passed / 0 failed / 1 ignored under `dex-osmosis`. Three new lifecycle tests landed: `open_lifecycle_happy_path`, `open_failed_on_error_refunds_and_finalises`, `late_open_lease_ack_after_open_failed_is_absorbed`.

## Off-chain consumer coordination (release-blocking)

- **`nolus-protocol/ibc-solray#331`** — Solana side must tolerate the relaxed `OpenLeaseParams` invariant (`downpayment == lpn` and `downpayment == asset`). Re-tighten on the Nolus side if Solana cannot handle one or both.
- **`nolus-protocol/ibc-solray#297`** — remote-profit migration still open; gates the eventual `swap` crate deletion.
- **Frontend `app.nolus.io`** — handle the new `OngoingTrx::RequestingOpenLease` + `OngoingTrx::OpenLease { remote_lease }` variants and the new `StateResponse::OpenFailed { reason }` top-level variant. `OngoingTrx::OpenIcaAccount` is removed.
- **Indexer / explorer** — subscribe to `wasm-ls-remote-lease-open-failed` and `wasm-ls-remote-lease-late-ack` events.
- **Deploy script (`scripts/deploy-protocol.sh:196`)** — must populate `expected_instance_ordinal` and `remote_lease_controller` on the leaser `InstantiateMsg`. Without these the leaser will fail to instantiate.

## Follow-ups (out of scope for this PR)

- Swap replacement (controller-driven `Lease::swap` replaces the in-lease DEX path).
- TransferOut / drain-home replacement.
- CloseLease state.
- Bounded retry + poison-callback terminal.
- Leaser v7 → v8 bump.
- Dead-code removal (`OpenIcaAccount`, `swap` crate, `dex::IcaConnector` from lease).
- Operator runbook for `Status::OpenFailed` (drafted; lives in local memory pending operator-docs landing).

### CODE REVIEW

File: protocol/packages/remote_lease_wire/src/callback.rs [hunk 1]
```diff
@@ -55,6 +59,21 @@ impl RemoteErrorMessage {
...
     }
 
+    /// Construct from a compile-time-known literal that the caller guarantees
+    /// is within [`OPERATION_ERR_MAX_BYTES`].
+    ///
+    /// For fixed internal reasons (e.g. `"timeout"`) where threading a
+    /// fallible [`new`](Self::new) through the call site would add error
+    /// plumbing for a value that is provably in range. The length invariant
+    /// is enforced with a `debug_assert!`; release builds trust the caller.
+    pub fn from_static(message: &'static str) -> Self {
```
Comment (kostovster) [inline protocol/packages/remote_lease_wire/src/callback.rs]: **[STYLE] CLAUDE.md Unit Test Guidelines 2 — test every public method** — `from_static` is a new public method with non-trivial debug-only precondition semantics but has no direct unit test in `remote_lease_wire`; it is only exercised indirectly through the lease contract integration test. Add a `#[test]` covering the within-cap success case (and ideally a `#[should_panic]` debug-build test at the boundary) alongside the existing `RemoteErrorMessage` tests.

File: protocol/contracts/leaser/src/state/leases.rs [hunk 2]
```diff
@@ -13,34 +15,97 @@ use crate::{
...
+        })
+    }
+
+    /// Read and remove the single in-flight open. Returns the customer +
+    /// state if present, `None` if nothing is pending. Singleton-flavoured
+    /// usage (only one open at a time) lives in the caller; the map shape
+    /// keeps all state per-customer.
+    fn take_pending(storage: &mut dyn Storage) -> ContractResult<Option<(Addr, PendingState)>> {
+        let first = Self::PENDING_OPENS
+            .range(storage, None, None, Order::Ascending)
```
[the remark this answers is not in the document]
Comment (gmanev7) [reply protocol/contracts/leaser/src/state/leases.rs]: Fixed in aaca055b5. `take_pending` no longer uses `range().next()`; the singleton `Item<PendingOpen>` carries the customer, and the instantiate reply consumes "the one open in flight" — it never correlates by `Addr` sort order. Added `test_remove_other_customer_keeps_pending` to pin that a removal for a different customer leaves the in-flight open intact.

File: protocol/packages/remote_lease_wire/src/callback.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (gmanev7) [reply protocol/packages/remote_lease_wire/src/callback.rs]: Resolving as working-as-intended, after clarifying the rule.

`from_static` is an internal, in-workspace API: its only non-test caller is the lease contract (`open_lease.rs`), passing fixed compile-time consts (`TIMEOUT_REASON` = 7 bytes, `UNEXPECTED_OPERATION_REASON` = 29 bytes). It is not a user/UI-facing surface.

We have tightened CLAUDE.md Code Style 23 to make the scope explicit: "public" / "external caller" means a surface reached directly by a user or UI (a contract `execute`/`query`/`migrate`/`sudo` entry point, its message/storage schemas, a CLI flag) — not a `pub` Rust item consumed by other code in the monorepo. In-workspace callers are trusted collaborators; an internal `pub fn` carries its contract through types, `debug_assert!`, and documentation, and is not required to defend at runtime against deliberate misuse (e.g. a `Box::leak`ed runtime string) by other workspace code.

The misuse-resistance bar therefore applies at the actual trust boundary — the counterparty-authored, wire-sourced `RemoteErrorMessage`. That path is hard-bounded at runtime: deserialization rejects over-cap payloads inside the visitor before allocating, and the fallible `new()` rejects them too. `from_static` exists only for fixed internal literals where threading a `Result` would add error plumbing for a value provably in range; its precondition is now documented with a `# Panics` section, and a debug-only unit test pins the over-cap panic.

File: protocol/contracts/leaser/src/state/leases.rs [hunk 3]
```diff
@@ -72,10 +168,35 @@ impl Leases {
...
+                    .save(
+                        storage,
+                        &PendingOpen {
+                            customer,
+                            phase: PendingPhase::Cancelled,
+                        },
+                    )
+                    .map(|()| true)
+                    .map_err(ContractError::RemoveLeaseFailure),
+                Some(_) | None => Ok(false),
```
[the remark this answers is not in the document]
Comment (gmanev7) [reply protocol/contracts/leaser/src/state/leases.rs]: Fixed in 52281ee84. The no-op arm now enumerates `Cached` (different customer), `Cancelled`, and `None` explicitly instead of `Some(_) | None`, so a new `PendingPhase` fails to compile rather than falling through silently. Behaviour is unchanged.

File: protocol/contracts/leaser/src/state/leases.rs [hunk 4]
```diff
@@ -161,6 +334,92 @@ mod test {
...
+        // The original customer's open is still `Cached` and registers.
+        assert_eq!(
+            SaveOutcome::Registered,
+            Leases::save(&mut storage, test_lease()).unwrap()
+        );
+        assert_lease_exist(&storage);
+    }
+
+    #[test]
+    fn test_cache_open_req_rejects_a_second_in_flight_open() {
```
[the remark this answers is not in the document]
Comment (gmanev7) [reply protocol/contracts/leaser/src/state/leases.rs]: Fixed in 52281ee84. Added `test_cache_open_req_reclaims_a_cancelled_slot`, which seeds a `Cancelled` slot (via `remove` after `cache_open_req`, asserting it returns `true` so the slot is provably `Cancelled` and not `None`) and asserts the next `cache_open_req` reclaims it and registers under the new customer.

File: protocol/packages/remote_lease_wire/src/callback.rs [hunk 5]
```diff
@@ -55,6 +59,32 @@ impl RemoteErrorMessage {
...
+    /// Precondition (caller's responsibility): `message` must be a genuine
+    /// literal whose length is verifiable by inspection, never a
+    /// runtime-produced `&'static str` (e.g. a `Box::leak`ed value). The
+    /// length is only `debug_assert!`ed, so an over-cap input that slips past
+    /// review would bypass the bound in release builds — unlike [`new`] and
+    /// deserialisation, which reject it. When the length is not statically
+    /// obvious, use [`new`] instead.
+    ///
+    /// # Panics
+    ///
```
[the remark this answers is not in the document]
Comment (gmanev7) [reply protocol/packages/remote_lease_wire/src/callback.rs]: Keeping the `debug_assert!`. Passing an over-cap `&'static str` to `from_static` is a contract violation (a programmer error), not runtime-reachable input — the only literals fed to it are crate-internal constants verifiable by inspection. The wire producer (the Solana side) constructs `RemoteErrorMessage` via the fallible `new` or via deserialization, both of which enforce `OPERATION_ERR_MAX_BYTES` unconditionally; `from_static` is public but not a user-facing surface. Per CLAUDE.md Code Style 23, internal code-to-code APIs state their contract through types and assertions and are not required to defend at runtime against deliberate misuse by other workspace code. Acknowledged; no change.
